### PR TITLE
Adds initial storage code.

### PR DESCRIPTION
**Pull request by mhoak**
Adds an initial storage file-store. I think this is a decent approach as it treats the storage as a pseudo database and should be able to grow as the needs evolve.

### CODE REVIEW

File: core/storage/diskstore.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+// Put is method to Put a file to disk
+func (ds *DiskStore) Put(t StoredDataType, filename string, json string) error {
+	path := ds.path(t, filename)
+	return ioutil.WriteFile(path, []byte(json), os.ModePerm)
+}
+
+func (ds *DiskStore) path(t StoredDataType, filename string) string {
+	return fmt.Sprintf("%s/%s/%s", ds.rootDir, t, filename)
```
Comment (artasparks) [inline core/storage/diskstore.go]: Prefer path.Join for system compatibility

File: core/storage/diskstore.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+func (ds *DiskStore) ensureDirectoryStructure(rootDir string) error {
+	fileInfo, err := os.Stat(rootDir)
+	if err != nil {
+		return err
+	}
+	if !fileInfo.IsDir() {
+		return errors.New("Directory does not exist")
+	}
+	for _, t := range storedDataTypes {
+		curDir := fmt.Sprintf("%s/%s", ds.rootDir, t)
```
Comment (artasparks) [inline core/storage/diskstore.go]: here also: path.Join

File: core/storage/diskstore.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		fileInfo, err := os.Stat(curDir)
+		if err != nil {
+			os.Mkdir(curDir, 0755)
```
Comment (artasparks) [inline core/storage/diskstore.go]: nit: Maybe somewhere having a 

```go
const defaultDirPrms = 0755
```

File: core/storage/diskstore.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+// ensureDirectoryStructure ensures that paths exist for each
+// of the expected file outputs
+func (ds *DiskStore) ensureDirectoryStructure(rootDir string) error {
+	fileInfo, err := os.Stat(rootDir)
+	if err != nil {
+		return err
+	}
+	if !fileInfo.IsDir() {
+		return errors.New("Directory does not exist")
```
Comment (artasparks) [inline core/storage/diskstore.go]: nit: errors usually are not capitalized: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: core/storage/diskstore_test.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	"io/ioutil"
+	"log"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+)
+
+func TestDiskGetNoFileErrors(t *testing.T) {
```
Comment (artasparks) [inline core/storage/diskstore_test.go:16]: nit: I get wary when I see real I/O in unit tests, even with tempdirs. I think it's better to mock out the I/O. That being said, this isn't a huge issue. Mocking out is a bit of work.

File: core/storage/filestore.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package storage
+
+// Filestore is a interface for file storage
+type Filestore interface {
```
Comment (artasparks) [inline core/storage/filestore.go:6]: Let's put `ctx context.Context` as a parameter to all the methods. We'll need that for all network Filestores.

File: core/storage/filestore.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+	Get(StoredDataType, string) (string, error)
+	List(StoredDataType, string) ([]string, error)
+	Put(StoredDataType, string, string) error
+}
+
+// StoredDataType designates which type of data is stored
+type StoredDataType string
+
+const (
+	// SGFInput storage directory
```
Comment (artasparks) [inline core/storage/filestore.go]: These names seem a little too wordy to me. I've provided some suggestions to abbreviate.

File: core/storage/filestore.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+	SGFInput StoredDataType = "input_sgf"
```
Comment (artasparks) [inline core/storage/filestore.go]: how about just `games`

File: core/storage/filestore.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	// AnalysisOutput is the data type for katago output
+	AnalysisOutput StoredDataType = "katago_analysis"
```
Comment (artasparks) [inline core/storage/filestore.go]: how about `analysis`?

File: core/storage/filestore.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+	// AiProblemIdentification is the directory for AI identifed problems
+	AiProblemIdentification StoredDataType = "ai_problem_identification"
```
Comment (artasparks) [inline core/storage/filestore.go]: I'm still not sure what this based on the name. Maybe `metadata`?

File: core/storage/filestore.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+	// AiProblemSGF is the directory for AI problems that have been generated
+	// and turned into SGF.
+	AiProblemSGF StoredDataType = "ai_problem_sgf"
```
Comment (artasparks) [inline core/storage/filestore.go]: How about just `problems`

File: core/storage/filestore.go [hunk 12]
```diff
@@ -0,0 +1,30 @@
+package storage
```
Comment (artasparks) [inline core/storage/filestore.go:1]: +package comment?

File: core/storage/diskstore_test.go (hunk 5 above)
Comment (mhoak) [reply core/storage/diskstore_test.go:16]: Hrm. I'm not sure if that's worth doing. If you want it done, I'm happy to do it.

File: core/storage/filestore.go (hunk 6 above)
Comment (mhoak) [reply core/storage/filestore.go:6]: I'm not familiar with this specific pattern in go. I've added it as I think you meant it. Let me know.

File: core/storage/diskstore.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+		rootDir: root,
+	}
+	err := ds.ensureDirectoryStructure(root)
+	if err != nil {
+		return nil, err
+	}
+	return ds, nil
+}
+
+// Get abstract method to get JSON content from a Filestore
```
Comment (artasparks) [inline core/storage/diskstore.go]: nit: content doesn't need to be JSON

File: core/storage/diskstore.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), nil
+}
+
+// List is method to list available files
+func (ds *DiskStore) List(ctx context.Context, t StoredDataType) ([]string, error) {
+	ioutil.ReadDir(ds.rootDir)
+	return []string{""}, nil
```
Comment (artasparks) [inline core/storage/diskstore.go]: Looks like an error here -- doesn't seem to be doing a thing.

File: core/storage/filestore.go (hunk 6 above)
Comment (artasparks) [reply core/storage/filestore.go:6]: This is right.

File: core/storage/filestore.go [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+type StoredDataType string
+
+const (
+	// Games storage directory
+	Games StoredDataType = "games"
+	// Analysis is the data type for katago output
+	Analysis StoredDataType = "analysis"
+	// Problems is the directory for AI problems that have been generated
+	// and turned into SGF.
+	Problems StoredDataType = "ai_problem_sgf"
```
Comment (artasparks) [inline core/storage/filestore.go]: maybe just gen_problems? Or even just problems/?

File: core/storage/diskstore_test.go (hunk 5 above)
Comment (artasparks) [reply core/storage/diskstore_test.go:16]: not a big deal

**Review by artasparks — APPROVED** [review]
LGTM
